src/data/actions/assets.js

```diff
@@ -121,46 +121,44 @@ export const uploadingAssets = count => ({
   count,
 });
 
-export const uploadAssetSuccess = (asset, response) => ({
+export const uploadAssetSuccess = response => ({
   type: assetActions.UPLOAD_ASSET_SUCCESS,
-  asset,
   response,
 });
 
-export const uploadAssetFailure = (file, response) => ({
+export const uploadAssetFailure = (asset, response) => ({
   type: assetActions.UPLOAD_ASSET_FAILURE,
-  file,
+  asset,
   response,
 });
 
 export const uploadAssets = (files, courseDetails) =>
   (dispatch, getState) => {
     dispatch(uploadingAssets(files.length));
-    files.forEach((file) => {
+    // gather all the promises into a single promise that can be returned
+    return Promise.all(files.map(file => (
       clientApi.postUploadAsset(courseDetails.id, file)
         .then((response) => {
           if (response.ok) {
-            response.json().then((data) => {
-              dispatch(uploadAssetSuccess(data.asset, response));
-              dispatch(getAssets(getState().request, courseDetails));
+            return response.json().then((json) => {
+              dispatch(uploadAssetSuccess(json));
+              // dispatch(getAssets(..)) returns a promise, so we return it
+              // up to chain it
+              return dispatch(getAssets(getState().request, courseDetails));
             });
-          } else {
-            dispatch(uploadAssetFailure(file, response));
           }
-        });
-    });
+          dispatch(uploadAssetFailure(file, response.status));
```

Comment:
I've changed `response` to `response.status` just because I don't want to send over the raw Response object. Is there a better alternative? Should I model this after the getAssets action creator, which creates an Error object using the Response and uses that as the action payload? See [here](https://github.com/edx/studio-frontend/blob/master/src/data/actions/assets.js#L35-L45).

Reply:
I think status is fine for now, once we have a valid use case of error responses (non-2XX) we can revaluate.